Tests/UnitTests/Core/TrackFitting/KalmanFitterTests.cpp

```diff
@@ -22,560 +20,483 @@
-      PropagatorPlainOptions(), rSurface);
-
-  KalmanFitter kFitter(rPropagator);
-
-  // Test 1: The nominal fit
-  auto fitRes = kFitter.fit(sourcelinks, rStart, kfOptions);
-  BOOST_CHECK(fitRes.ok());
-  auto& fittedTrack = *fitRes;
-  auto fittedParameters = fittedTrack.fittedParameters.value();
+BOOST_AUTO_TEST_CASE(ZeroFieldWithHole) {
+  auto start = makeParameters();
+  auto measurements = createMeasurements(simPropagator, geoCtx, magCtx, start,
+                                         resolutions, rng);
+  const auto& sourceLinks = measurements.sourceLinks;
+  BOOST_REQUIRE_EQUAL(sourceLinks.size(), nMeasurements);
+
+  // fitter options w/o target surface
+  KalmanFitterOptions<TestSourceLinkCalibrator, VoidOutlierFinder> kfOptions(
+      geoCtx, magCtx, calCtx, TestSourceLinkCalibrator(), VoidOutlierFinder(),
+      LoggerWrapper{*kfLogger}, PropagatorPlainOptions());
+
+  // always keep the first and last measurement. leaving those in seems to not
+  // count the respective surfaces as holes.
+  for (size_t i = 1u; (i + 1u) < sourceLinks.size(); ++i) {
+    // remove the i-th measurement
+    auto withHole = sourceLinks;
+    withHole.erase(std::next(withHole.begin(), i));
+    BOOST_REQUIRE_EQUAL(withHole.size() + 1u, sourceLinks.size());
+    BOOST_TEST_INFO("Removed measurement " << i);
+
+    auto res = kfZero.fit(withHole, start, kfOptions);
+    BOOST_REQUIRE(res.ok());
+
+    const auto& val = res.value();
+    BOOST_CHECK_NE(val.trackTip, SIZE_MAX);
+    BOOST_CHECK(not val.fittedParameters);
+    BOOST_CHECK_EQUAL(val.measurementStates, withHole.size());
+    // check the output status flags
+    BOOST_CHECK(val.smoothed);
+    BOOST_CHECK(not val.reversed);
+    BOOST_CHECK(not val.reset);
+    BOOST_CHECK(val.finished);
+    BOOST_CHECK_EQUAL(val.missedActiveSurfaces.size(), 1u);
+  }
+}
 
-  // Calculate global track parameters covariance matrix
-  const auto& [trackParamsCov, stateRowIndices] =
-      detail::globalTrackParametersCovariance(fittedTrack.fittedStates,
-                                              fittedTrack.trackTip);
-
-  // Check the size of the global track parameters size
-  BOOST_CHECK_EQUAL(stateRowIndices.size(), 6);
-  BOOST_CHECK_EQUAL(stateRowIndices.at(fittedTrack.trackTip), 30);
-  BOOST_CHECK_EQUAL(trackParamsCov.rows(), 6 * eBoundSize);
-
-  // Test 2: Make sure it is deterministic
-  fitRes = kFitter.fit(sourcelinks, rStart, kfOptions);
-  BOOST_CHECK(fitRes.ok());
-  auto& fittedAgainTrack = *fitRes;
-  auto fittedAgainParameters = fittedAgainTrack.fittedParameters.value();
-
-  CHECK_CLOSE_REL(fittedParameters.parameters().template head<5>(),
-                  fittedAgainParameters.parameters().template head<5>(), 1e-5);
-  CHECK_CLOSE_ABS(fittedParameters.parameters().template tail<1>(),
-                  fittedAgainParameters.parameters().template tail<1>(), 1e-5);
-
-  // Test 3: Fit without target surface
-  kfOptions.referenceSurface = nullptr;
```

Comment:
Do you have a test for fitting without target surface as well?

Reply:
The first test `ZeroField` runs the default configuration without a target surface.

Reply:
I see. Great. Thanks!

Reply:
I re-added the missing test but it fails w/ a smoothing error.